integration-tests/src/tests/client/features/delegate_action.rs

```diff
@@ -287,6 +291,172 @@ fn meta_tx_fn_call() {
+
+    let method_name = "log_something".to_owned();
+    let method_name_len = method_name.len() as u64;
+    // 1 yocto near, that's less than 1 gas unit
+    let initial_allowance = 1;
+    let signer = create_user_test_signer(&sender);
+    let access_key = AccessKey {
+        nonce: 0,
+        permission: AccessKeyPermission::FunctionCall(FunctionCallPermission {
+            allowance: Some(initial_allowance),
+            receiver_id: receiver.to_string(),
+            method_names: vec![method_name.clone()],
+        }),
+    };
+
+    let mut genesis = Genesis::test(vec![relayer.clone(), receiver.clone()], 3);
+    add_test_contract(&mut genesis, &receiver);
+    add_account_with_access_key(&mut genesis, &sender, NEAR_BASE, signer.public_key(), access_key);
+    let node = RuntimeNode::new_from_genesis(&relayer, genesis);
+
+    let actions = vec![Action::FunctionCall(FunctionCallAction {
+        method_name,
+        args: vec![],
+        gas: 30_000_000_000_000,
+        deposit: 0,
+    })];
+
+    // this should still succeed because we use the gas of the relayer, not of the access key
+    let outcome =
+        check_meta_tx_fn_call(&node, actions, method_name_len, 0, sender, relayer, receiver);
+
+    // Check that the function call was executed as expected
+    let fn_call_logs = &outcome.receipts_outcome[1].outcome.logs;
+    assert_eq!(fn_call_logs, &vec!["hello".to_owned()]);
+}
+
+/// Call a function in a meta tx where the user doesn't have the appropriate
+/// access key, which must fail.
+///
+/// This is quite to fail, method restricted access keys can give restricted
+/// access to a contract. If meta transactions can be used to circumvent this
+/// check, then someone with an access key could impersonate the account in
+/// unintended ways.
+#[test]
+fn meta_tx_fn_call_access_wrong_method() {
+    let sender = bob_account();
+    let relayer = alice_account();
+    let receiver = carol_account();
+
+    let method_name = "log_something".to_owned();
+    let access_key_method_name = "log_something_else".to_owned();
+    let signer = create_user_test_signer(&sender);
+    let access_key = AccessKey {
+        nonce: 0,
+        permission: AccessKeyPermission::FunctionCall(FunctionCallPermission {
+            allowance: Some(NEAR_BASE),
+            receiver_id: receiver.to_string(),
+            method_names: vec![access_key_method_name],
+        }),
+    };
+
+    let mut genesis = Genesis::test(vec![relayer.clone(), receiver.clone()], 3);
+    add_test_contract(&mut genesis, &receiver);
+    add_account_with_access_key(&mut genesis, &sender, NEAR_BASE, signer.public_key(), access_key);
+    let node = RuntimeNode::new_from_genesis(&relayer, genesis);
+
+    let actions = vec![Action::FunctionCall(FunctionCallAction {
+        method_name,
+        args: vec![],
+        gas: 30_000_000_000_000,
+        deposit: 0,
+    })];
+
+    let tx_result = node.user().meta_tx(sender, receiver, relayer, actions).unwrap();
```

Comment:
There's an unfortunate amount of duplication between the three tests added here? Would it be possible to consolidate them somehow?

Reply:
I tried my best to reduce duplication, but the consolidated function need an awful lot of parameters. See my latest commit.

In terms of LOC, I ended up with more code than before. But I think it's still an improvement, as the logical duplication is reduced. Not sure if the code is cleaner to read though.